From 19ea45f119475a7f8f124f63e02a8ea7283f7efd Mon Sep 17 00:00:00 2001
From: Sakura286 <Sakura286>
Date: Fri, 22 Aug 2025 04:07:03 +0000
Subject: [PATCH 04/12] enlarge extra_hw_frames for h26x

---
 dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
index 176ae05a48..4135a8e6b9 100644
--- a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
+++ b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
@@ -1060,11 +1060,19 @@ void FFmpegVideoDecoder<LIBAV_VER>::InitHWCodecContext(ContextType aType) {
   }
 
   if (mCodecID == AV_CODEC_ID_H264) {
+#if !defined(ES_MPP)
     mCodecContext->extra_hw_frames =
         H264::ComputeMaxRefFrames(mInfo.mExtraData);
+#else
+    mCodecContext->extra_hw_frames = 16;
+#endif
   } else if (mCodecID == AV_CODEC_ID_HEVC) {
+#if !defined(ES_MPP)
     mCodecContext->extra_hw_frames =
         H265::ComputeMaxRefFrames(mInfo.mExtraData);
+#else
+    mCodecContext->extra_hw_frames = 16;
+#endif
   } else {
     mCodecContext->extra_hw_frames = EXTRA_HW_FRAMES;
   }
-- 
2.43.0

