From d7cb7e38e307b61f7bc7df18cc3f19b8d5f1da32 Mon Sep 17 00:00:00 2001
From: CHEN <chenxuan@iscas.ac.cn>
Date: Thu, 16 Oct 2025 09:55:19 +0000
Subject: [PATCH 04/12] enlarge extra_hw_frames for h26x

---
 dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
index f9d4256aec..498e1e3e22 100644
--- a/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
+++ b/dom/media/platforms/ffmpeg/FFmpegVideoDecoder.cpp
@@ -1125,11 +1125,19 @@ void FFmpegVideoDecoder<LIBAV_VER>::InitHWCodecContext(ContextType aType) {
   }
 
   if (mCodecID == AV_CODEC_ID_H264) {
+#if !defined(ES_MPP)
     mCodecContext->extra_hw_frames =
         H264::ComputeMaxRefFrames(mInfo.mExtraData);
+#else
+    mCodecContext->extra_hw_frames = 16;
+#endif
   } else if (mCodecID == AV_CODEC_ID_HEVC) {
+#if !defined(ES_MPP)
     mCodecContext->extra_hw_frames =
         H265::ComputeMaxRefFrames(mInfo.mExtraData);
+#else
+    mCodecContext->extra_hw_frames = 16;
+#endif
   } else {
     mCodecContext->extra_hw_frames = EXTRA_HW_FRAMES;
   }
-- 
2.48.1

