From 5a8c94172431d1690ab854bdecd26d6f73204ad5 Mon Sep 17 00:00:00 2001
From: CHEN Xuan <chenxuan@iscas.ac.cn>
Date: Tue, 1 Jul 2025 10:52:12 +0000
Subject: [PATCH 08/12] stop blocking some devices from vaapi accelerate

---
 widget/gtk/GfxInfo.cpp | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/widget/gtk/GfxInfo.cpp b/widget/gtk/GfxInfo.cpp
index 4589e920a7..c7234c9ffa 100644
--- a/widget/gtk/GfxInfo.cpp
+++ b/widget/gtk/GfxInfo.cpp
@@ -1168,12 +1168,12 @@ const nsTArray<GfxDriverInfo>& GfxInfo::GetGfxDriverInfo() {
         V(0, 0, 0, 0), "FEATURE_HARDWARE_VIDEO_DECODING_NO_LINUX_AMD", "");
 
     // Disable on r600 driver due to decoding artifacts (Bug 1824307)
-    APPEND_TO_DRIVER_BLOCKLIST_EXT(
-        OperatingSystem::Linux, ScreenSizeStatus::All, BatteryStatus::All,
-        WindowProtocol::All, DriverVendor::MesaR600, DeviceFamily::All,
-        nsIGfxInfo::FEATURE_HARDWARE_VIDEO_DECODING,
-        nsIGfxInfo::FEATURE_BLOCKED_DEVICE, DRIVER_COMPARISON_IGNORED,
-        V(0, 0, 0, 0), "FEATURE_HARDWARE_VIDEO_DECODING_NO_R600", "");
+    // APPEND_TO_DRIVER_BLOCKLIST_EXT(
+    //     OperatingSystem::Linux, ScreenSizeStatus::All, BatteryStatus::All,
+    //     WindowProtocol::All, DriverVendor::MesaR600, DeviceFamily::All,
+    //     nsIGfxInfo::FEATURE_HARDWARE_VIDEO_DECODING,
+    //     nsIGfxInfo::FEATURE_BLOCKED_DEVICE, DRIVER_COMPARISON_IGNORED,
+    //     V(0, 0, 0, 0), "FEATURE_HARDWARE_VIDEO_DECODING_NO_R600", "");
 
     // Disable on AMD devices using broken Mesa (Bug 1837140).
     // https://gitlab.freedesktop.org/mesa/mesa/-/commit/0c024bbe641b092bbb
-- 
2.43.0

